internal/controller/certificate_controller.go

```
@@ -0,0 +1,157 @@
+		return ctrl.Result{}, fmt.Errorf(errCreationFailed, err)
+	}
+
+	if isCertificateValid(certificate, certificateConfig) {
+		r.removeErrorConditions(ctx, certificate)
+		return ctrl.Result{}, nil
+	}
+
+	validTo, validFrom, signatureHashAlgorithm, condition, err := r.issueCertificate(ctx, certificateConfig, certificate)
+	if err != nil {
+		r.updateCertificateConditions(ctx, certificate, condition)
+		return ctrl.Result{}, err
+	}
+
+	condition, err = r.updateCertValidity(ctx, certificate, validTo, validFrom, signatureHashAlgorithm)
+	if err != nil {
+		r.updateCertificateConditions(ctx, certificate, condition)
+		return ctrl.Result{}, err
+	}
+
+	tlsData, condition, err := r.downloadCert(ctx, certificate, certificateConfig)
+	if err != nil {
+		r.updateCertificateConditions(ctx, certificate, condition)
+		return ctrl.Result{}, err
+	}
+
+	condition, err = r.createOrUpdateTlsSecret(ctx, certificate, tlsData, req.Namespace)
+	if err != nil {
+		r.updateCertificateConditions(ctx, certificate, condition)
+		return ctrl.Result{}, err
+	}
+
+	r.removeErrorConditions(ctx, certificate)
+	return reconcile.Result{}, nil
+}
+
+// updateCertificateConditions updates the conditions of the Certificate resource
+func (r *CertificateReconciler) updateCertificateConditions(ctx context.Context, certificate *v1alpha1.Certificate, condition metav1.Condition) {
+	meta.SetStatusCondition(&certificate.Status.Conditions, condition)
+	err := r.Client.Status().Update(ctx, certificate)
+	if err != nil {
+		r.Log.Error(err, errUpdateStatus)
+	}
+}
+
+// removeErrorConditions removes the error conditions of the Certificate resource
+func (r *CertificateReconciler) removeErrorConditions(ctx context.Context, certificate *v1alpha1.Certificate) {
+	meta.RemoveStatusCondition(&certificate.Status.Conditions, ConditionError)
+
+	err := r.Client.Status().Update(ctx, certificate)
+	if err != nil {
+		r.Log.Error(err, errUpdateStatus)
+	}
+}
```

I'm not sure about how I feel about these errors being unhandled apart from logging the error. I think we should return an error normally here and handle it at the caller. Here's a similar [example](https://github.com/dana-team/hns/blob/e8246b4a73e9efa5d2e4875d8a2bba9f79b7ab30/internal/updatequota/controller.go#L92-L98) from a different project. It kind of clatters the Reconciler and makes it less "pretty" but probably needed. WDYT?